kernel/ls3_fs.c

```diff
@@ -0,0 +1,1058 @@
+    pr_info("Mounting device %s", (char*)data);
+    struct file *filp = filp_open((char*)data, O_RDWR, 0644);
+
+    struct superblock *super = kzalloc(sizeof(struct superblock), GFP_USER);
+
+    // Read the superblock
+    loff_t pos = 0; // FIXME: should scan first 32 blocks to find most recent
+    kernel_read(filp, super, PAGE_CACHE_SIZE, &pos);
+
+    // Sanity checks
+    if (super->magic != LS3_MAGIC) {
+        pr_err("BAD MAGIC NUMBER");
+        return -EIO;
+    }
+    // When mounting, i_read_size(/dev/loop0) returns 0, so we can't
+    // do this check...
+    // if (super->fs_size != (uint64_t)i_size_read(file_inode(filp))) {
+    //     pr_err("ERROR: superblock fs_size (%lld) does not match device size (%lld)\n",
+    //             super->fs_size, i_size_read(file_inode(filp)));
+    //     return -EBADF;
+    // }
+
+    // Store in mount table
+    struct mount_data *m = alloc_mount(super, filp);
+    mnt[mountno] = m;
+    pr_info("Recovered mount structure:\n"
+            "             super->fs_size = %lld\n"
+            "                 num_blocks = %lld\n"
+            "          num_freemap_bytes = %lld\n"
+            "         num_freemap_blocks = %lld\n"
+            "   num_total_freemap_blocks = %lld\n",
+            super->fs_size, m->num_blocks, m->num_freemap_bytes, m->num_freemap_blocks, m->num_total_freemap_blocks);
+    for(i = 0; i < m->num_freemap_blocks; i++) {
+        pr_info("           freemap_block[%d] = %lld", i, m->super->freemap_block[i]);
+    }
+
+    // Load freemap from device, and
+    // calculate highest_freemap_blockno
+    for(i = 0; i < m->num_freemap_blocks; i++) {
+        uint64_t blockno = mnt[mountno]->super->freemap_block[i];
+        if (blockno > m->highest_freemap_blockno)
+            m->highest_freemap_blockno = blockno;
+        get_block(blockno, ((void *)mnt[mountno]->freemap)+(i*PAGE_CACHE_SIZE), mountno);
+    }
+
+    if (verbose >= 2) {
+        pr_info("freemap:\n");
+        dump_bitmap(m->freemap, m->num_blocks);
+    }
+
+    if (verbose >= 3) {
+        dump_all_blocks_hex(mountno, verbose == 3);
+    }
+
+    struct inode *root;
+    struct dentry *root_dentry;
+
+    sb->s_blocksize = PAGE_CACHE_SIZE;
+    sb->s_blocksize_bits = PAGE_CACHE_SHIFT;
+    sb->s_magic = LS3_MAGIC;
+    sb->s_op = &ls3_s_ops;
+    // superblock private info is just an integer, which we use
+    // to index into a static array of mount_data structures.
+    sb->s_fs_info = kzalloc(sizeof(int), GFP_USER);
+    *(int *)sb->s_fs_info = mountno;
+
+    //sb->s_maxbytes  = MAX_LFS_FILESIZE;
+
+    root = ls3_make_inode(sb, S_IFDIR | 0755);
+    if (! root)
+        goto out;
+    root->i_op = &simple_dir_inode_operations;
+    root->i_fop = &simple_dir_operations;
+
+    root_dentry = d_make_root(root);
+    if (! root_dentry)
+        goto out_iput;
+    sb->s_root = root_dentry;
+
+    //ls3_create_files (sb, root_dentry);
+    pr_info("finished mounting %d", mountno);
+    return 0;
```

Comment:
this function looks similar enough to mine but I think this is the issue somehow